turbopack/crates/turbo-tasks/src/effect.rs

```diff
@@ -0,0 +1,109 @@
+use std::{borrow::Cow, future::Future, panic, pin::Pin};
+
+use anyhow::{anyhow, Result};
+use auto_hash_map::AutoSet;
+use parking_lot::Mutex;
+use tokio::task::JoinHandle;
+use tracing::{Instrument, Span};
+
+use crate::{self as turbo_tasks, emit, manager::turbo_tasks_future_scope, CollectiblesSource, Vc};
+
+#[turbo_tasks::value_trait]
+trait Effect {}
+
+type EffectFuture = Pin<Box<dyn Future<Output = Result<()>> + Send + Sync + 'static>>;
+
+struct EffectInner {
+    future: EffectFuture,
+    span: Span,
+}
+
+#[turbo_tasks::value(serialization = "none", cell = "new", eq = "manual")]
+struct EffectInstance {
+    #[turbo_tasks(trace_ignore, debug_ignore)]
+    inner: Mutex<Option<EffectInner>>,
+}
+
+impl EffectInstance {
+    fn new(future: impl Future<Output = Result<()>> + Send + Sync + 'static) -> Self {
+        Self {
+            inner: Mutex::new(Some(EffectInner {
+                future: Box::pin(future),
+                span: Span::current(),
+            })),
+        }
+    }
+
+    pub fn apply(&self) -> Option<JoinHandle<Result<()>>> {
+        let future = self.inner.lock().take();
+        future.map(|EffectInner { future, span }| {
+            tokio::spawn(
+                turbo_tasks_future_scope(turbo_tasks::turbo_tasks(), future).instrument(span),
+            )
+        })
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl Effect for EffectInstance {}
+
+pub fn effect(future: impl Future<Output = Result<()>> + Send + Sync + 'static) {
+    emit::<Box<dyn Effect>>(Vc::upcast(EffectInstance::new(future).cell()));
+}
+
+pub async fn apply_effects(source: impl CollectiblesSource) -> Result<()> {
+    let effects: AutoSet<Vc<Box<dyn Effect>>> = source.take_collectibles();
+    if effects.is_empty() {
+        return Ok(());
+    }
+    let span = tracing::span!(tracing::Level::INFO, "apply effects", count = effects.len());
+    async move {
+        let mut first_error = anyhow::Ok(());
+        for effect in effects {
+            let Some(effect) = Vc::try_resolve_downcast_type::<EffectInstance>(effect).await?
+            else {
+                panic!("Effect must only be implemented by EffectInstance");
```

Comment:
You can better enforce this using the sealed trait pattern:

- https://predr.ag/blog/definitive-guide-to-sealed-traits-in-rust/#sealing-traits-with-a-supertrait
- https://rust-lang.github.io/api-guidelines/future-proofing.html#c-sealed

But please document that it's using the sealed trait pattern!

Reply:
The trait is private, do I still need to use this pattern.

Reply:
Oh, I guess not. 👍